### PR TITLE
driver: uart: ns16550: manual CTS check in poll_out()

### PR DESCRIPTION
**Pull request by fabled**
This implements manual hardware flow control for polled output, for the NS16550 variants that need it (e.g. do not have automatic hardware flow control support on chip level).

Suggestions on how to add the hardware flow control to polled input, and/or irq/fifo mode would be welcome.

I suppose for `poll_in()` we just turn off RTS# in the initialization so that no data is received. And, then `poll_in()` is modified to first drain bytes from the hardware fifo, and finally if fifo is empty, to enable RTS# until there is at least one byte in the fifo that can be read and returned, and then again turn RTS# off. That is RTS# would be logically only when in `poll_in()`. This can cause some performance impact, but probably the only way to do it. So with hardware flow control it would be preferable to use interrupt mode.

For the interrupt mode, I think RTS# could be driven manually based on the hardware RX fifo interrupts (assert when fifo empty, deassert when fifo "almost full" interrupt comes). For the TX side, the we probably need to check CTS# when filling TX fifo, and if it's asserted we go to 'stalled' state and stop feeding data to the FIFO. Then we wait for the CTS interrupt, and once it comes we go back to allowing data being sent to the FIFO. So basically emulate "fifo full" condition when CTS# is asserted, until the CTS# deasserted interrupt comes.

Would the above approach sound as a viable plan? Is there perhaps any other uart driver with similar CTS/RTS handling in the driver?

### CODE REVIEW

**Comment by fabled** [comment]
The build failures are from `drivers/can/can_mcux_flexcan.c` and from the current git master. This PR is not introducing the failures reported in Shippable.

**Comment by dcpleung** [comment]
What would happen if CTS line is not connected? Would this patch prevent us from only using RX/TX lines?

**Comment by fabled** [comment]
> What would happen if CTS line is not connected? Would this patch prevent us from only using RX/TX lines?

The check is done only if CTS/RTS has been reqested (via DTS tag, or explicit port configuration call). Perhaps this should have been noted in the commit message.

In any case the default is no hardware flow control. But if it is explicitly configured on, and the wires are not connected, things will no work, but it is user configuration issue.

**Review by dcpleung — APPROVED** [review]

**Comment by jenmwms** [comment]
> Perhaps this should have been noted in the commit message.

Yes, please include the clarification in the commit message, or might be better captured in the source code. Also, it would be great to see in the commit message a note on how you tested this patch.

**Comment by fabled** [comment]
>  Also, it would be great to see in the commit message a note on how you tested this patch.

Sure.

> Suggestions on how to add the hardware flow control to polled input, and/or irq/fifo mode would be welcome.

Any ideas for the above, or comments on the potential approaches listed? I am thinking if I should implement he polled input in the same patch. Or do you think all of these would be worthy of separate commits?

**Comment by fabled** [comment]
Commit message updated as requested.

**Comment by jenmwms** [comment]
> > Suggestions on how to add the hardware flow control to polled input, and/or irq/fifo mode would be welcome.
> 
> Any ideas for the above, or comments on the potential approaches listed? I am thinking if I should implement he polled input in the same patch. Or do you think all of these would be worthy of separate commits?

I cannot offer suggestions right now (need to look at it more), but it sounds reasonable to open an RFC or PR for the remaining work/discussion. You can link/mention them to track it. 

Any feedback others can offer? Any objections?

File: drivers/serial/uart_ns16550.c [hunk 1]
```diff
@@ -505,8 +505,21 @@ static int uart_ns16550_poll_in(struct device *dev, unsigned char *c)
 static void uart_ns16550_poll_out(struct device *dev,
 					   unsigned char c)
 {
+	struct uart_ns16550_dev_data_t *data = DEV_DATA(dev);
+
 	/* wait for transmitter to ready to accept a character */
-	while ((INBYTE(LSR(dev)) & LSR_THRE) == 0) {
+	while (1) {
```
Comment (jenmwms) [inline drivers/serial/uart_ns16550.c:511]: I think you need to rebase. When you do, please note the changes in commit 1cd5578 `serial: ns16550: Simplify poll out` . The `while (1)` was removed in favor of the line 509 you see here. Will you need to bring it back?

Comment (fabled) [reply drivers/serial/uart_ns16550.c:542]: Rebased. Seems reverting previous commit was the most sensible way to do this change and keep the code readable.

**Comment by jenmwms** [comment]
> This implements manual hardware flow control for polled output, for the NS16550 variants that need it (e.g. do not have automatic hardware flow control support on chip level).

Can you please give an example platform?  Are you able to test this on HW?

**Comment by fabled** [comment]
> Can you please give an example platform? Are you able to test this on HW?

I did add to the commit message the note:
```
Tested on CEC1702 NS16550 connected to FT230X USB UART
and verified that Zephyr will throttle output based on CTS.
```

Though, it seems I need to look at what fails in the test suite now after the rebase.

**Comment by fabled** [comment]
Updated with draft support for TX interrupt mode and RX paths. Please take a look if the concept looks good. I am going to test it locally next.

**Pull request closed without merging**